google-sheets/functions/log-sms.js

```diff
@@ -0,0 +1,32 @@
+const { google } = require('googleapis');
+
+exports.handler = async function(context, event, callback) {
+  const auth = new google.auth.GoogleAuth({
+    keyFile: context.SHEETS_SERVICE_ACCOUNT_JSON,
+    scopes: ['https://www.googleapis.com/auth/spreadsheets'],
+  });
+  const twiml = new Twilio.twiml.MessagingResponse();
+
+  try {
+    const sheets = google.sheets({
+      version: 'v4',
+      auth,
+    });
+
+    const res = await sheets.spreadsheets.values.append({
+      spreadsheetId: context.SHEETS_DOC_ID,
```

Comment:
If the doc id or sheet name are invalid, what kind of error does this produce?

Is there a way we might be able to build into the index.html a status check to ensure the data provided, auth, doc id, sheet name are all valid? Otherwise my fear is this will break and not help provide any context to the user as to why

Reply:
For an invalid or empty document ID, line 16 throws an error object with `error.code === 404`. For an invalid sheet name, we get a more general `error.code === 400` with a somewhat descriptive error message. I included some logic in the template to log more helpful errors for these cases to the debugger.

Do we have an example in `function-templates` of an index.html that does status checks? My first thought is to write a function that checks the status codes of calls to the Google API with the same client and credentials as the existing function and returns status JSON to the browser, but I want to make sure this is something I get right because the authentication credentials and data involved are potentially sensitive.

Reply:
My concern here when this is in a quick deploy environment, i'm not 100% sure if the twilio console debugger captures that for the user automatically, though if it does, the user might not be even aware to check the console or what to do about the error they see.

The only app I can think of that might have done any UI validation is for the Verify app. But that one is a more full featured experience, and not just the one-off index pages we have been building.

I think it would be worth making a small status check, and pointing the user to the `Edit This Code` button that gets rendered from the `APP_INFO` comment in the index page.

Due to the sensitive nature of the data of the auth token, if we rename the file `log-sms.private.js` it won't be publicly accessible

Reply:
If this needs to be a webhook it should be protected since Twilio won't be able to call private endpoints. I agree that we should avoid making endpoints public if we don't have to